martian/src/metadata.rs

```diff
@@ -146,67 +153,73 @@ impl Metadata {
             metadata_path,
             files_path,
             run_file,
             raw_jobinfo: Map::new(),
             jobinfo: Default::default(),
-            alarm_file,
-        }
+        })
     }
 
     /// Path within chunk
     pub fn make_path(&self, name: &str) -> PathBuf {
         make_metadata_file_path(self.metadata_path.as_ref(), name)
     }
 
-    /// Write to a file inside the chunk
-    pub fn write_raw(&mut self, name: &str, text: &str) -> Result<()> {
-        let mut f = File::create(self.make_path(name))?;
-        f.write_all(text.as_bytes())?;
-        // Ensure the file is closed before we write the journal, to reduce
-        // the chances that `mrp` sees the journal entry before the file content
-        // has be sync'ed.  This can be an issue on nfs systems.
+    /// Append to a metadata file inside the chunk.
+    ///
+    /// This operation is not atomic at the file system level.
+    pub(crate) fn append_metadata(&mut self, name: &str, content: &str) -> Result<()> {
+        let mut f = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(self.make_path(name))?;
+        f.write_all(content.as_bytes())?;
+        f.sync_all()?;
         drop(f);
+        self.update_journal(name)
+    }
+
+    /// Write to a metadata file, overwriting existing contents.
+    ///
+    /// The file write is fully atomic.
+    pub(crate) fn write_metadata(&mut self, name: &str, content: &str) -> Result<()> {
+        fully_atomic_write(&self.make_path(name), |w| w.write_all(content.as_bytes()))?;
         self.update_journal(name)?;
         Ok(())
     }
 
     /// Update the Martian journal -- so that Martian knows what we've updated
     fn update_journal(&self, name: &str) -> Result<()> {
-        let journal_name: Cow<'_, str> = if self.stage_type != "main" {
+        let journal_name: Cow<'_, str> = if self.stage_type != StageType::Main {
             format!("{}_{name}", self.stage_type).into()
         } else {
             name.into()
         };
-
-        let tmp_run_file = format!("{}.{journal_name}.tmp", self.run_file);
-        let run_file = &tmp_run_file[..tmp_run_file.len() - 4];
-
-        {
-            let mut f = File::create(&tmp_run_file)?;
-            if let Err(err) = f.write_all(make_timestamp_now().as_bytes()) {
-                // Pretty much ignore this error.  The only reason we need
-                // any content at all in this file is because some
-                // filesystems behave strangely with completely empty files.
-                eprintln!("Writing journal file {tmp_run_file}: {err}");
-            }
+        let journal_file = format!("{}.{journal_name}", self.run_file);
+        let timestamp = make_timestamp_now();
+        let mut write_err = Ok(());
+        let create_result = fully_atomic_write(Path::new(&journal_file), |w| {
+            write_err = w.write_all(timestamp.as_bytes());
+            Ok(())
+        });
```

Comment:
This should probably not be using `fully_atomic_write`.  Trying to make the write to the journal file atomic is a waste of effort.